### PR TITLE
ISSUE4 - Checks for duplicate SampleIDs

### LINKED ISSUE #4: Check for duplicate samples and mismatch between counts/metadata
perform checks like - 
- no duplicate samples in counts or metadata 
- all samples in counts match samples in metadata 
- all samples in metadata match samples in counts

### PR DESCRIPTION
**Pull request by daisyhan97**
Fixes #4 
Adds function to check for duplicate sample IDs in metadata/counts matrix, then ensures each sample in counts matrix has corresponding metadata and vice versa.

### CODE REVIEW

File: tests/testthat/test-check_mismatch.R [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  sample(0:100, size = 16),
+  ncol = 4,
+  dimnames = list(c("ENSG00000229807.12", "ENSG00000183878.12",
+                    "ENSG00000239807.12", "ENSG00000259807.12"),
+                  c("S1", "S2", "S3", "S3"))
+))
+
+test_that("errors if duplicate sampleids in counts matrix", {
+  expect_error(check_mismatch(metadata, counts))
+})
```
Comment (kelshmo) [inline tests/testthat/test-check_mismatch.R:79]:
```suggestion
})

```
small syntax suggestion, GH wants a newline.

File: R/functions.R [hunk 2]
```diff
@@ -242,6 +242,34 @@ get_biomart <- function(count_df, synid, version, host, filters, organism) {
...
+#'@inheritParams get_biomart
+#'@param md A data frame with sample identifiers as row names and relevant experimental covariates.
+#'@export
+check_mismatch <- function(md, count_df) {
+  # Isolate the SampleID's from the metadata and counts matrix
+  md_sampleid <- rownames(md)
+  cnt_sampleid <- colnames(count_df)
+  
+  # First check if these are duplicates in either of these. Returns an error if duplicates are found.
+  if (length(unique(md_sampleid)) != length(md_sampleid) ){
```
Comment (kelshmo) [inline R/functions.R:262]: We can modify this conditional since we inherently know duplicate row names are not allowed. Instead of checking for duplicates, let's start right away with checking the output of `setdiff` as you have it on line 268.

File: R/functions.R [hunk 3]
```diff
@@ -242,6 +242,34 @@ get_biomart <- function(count_df, synid, version, host, filters, organism) {
...
+    stop("ERROR: Unique sample IDs are required for subsequent functions. Please check your metadata input.")}
+  else if (length(unique(cnt_sampleid)) != length(cnt_sampleid)){
+    stop("ERROR: Unique sample IDs are required for subsequent functions.Please check your counts input")}
+  
+  # Ensures each sample in the metadata has a corresponding column in the counts matrix and vice versa
+  else if ((length(setdiff(md_sampleid, cnt_sampleid)) > 0) || (length(setdiff(cnt_sampleid, md_sampleid)) > 0)){
+    stop("ERROR: All SampleIDs provided must exist both in the counts and metadata inputs")
```
Comment (kelshmo) [inline R/functions.R:269]: I think it could be clearer to say: "All sample identifiers in the counts matrix and metadata file must match."

Also, you don't need to type "ERROR" as `stop()` will append it automatically: 

``` r
x = 3

if (x > 1) {stop("stop function")}
#> Error in eval(expr, envir, enclos): stop function
```

<sup>Created on 2021-07-30 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Review by kelshmo — CHANGES_REQUESTED** [review]
@daisyhan97 Thanks for the PR! This looks great!

My fault - since I suggested adding a test for duplicates. I forgot that setting duplicate rownames is not possible! That is why your test is failing at line 42. 

Since we use `tibble::column_to_rownames()` earlier in the plan, we can expect the user to receive a pretty prescriptive error message. If they tried to pass duplicate sample names in the metadata file, they would return this error:
``` r
metadata <- data.frame(
  batch = c("1", "2", "1", "2"),
  diagnosis = c("dx", "dx", "ct", "ct"),
  samples = c("S1", "S2", "S3", "S3"),
  stringsAsFactors = TRUE
)

tibble::column_to_rownames(metadata, var = "samples")
#> Warning: non-unique value when setting 'row.names': 'S3'
#> Error in `.rowNamesDF<-`(x, value = value): duplicate 'row.names' are not allowed
```

<sup>Created on 2021-07-30 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

And if they tried to pass duplicate sample names in the counts file, the columns would automatically rename themselves by appending a numeric value (see below), and then our check would alert them to mismatched identifiers between counts and metadata. 
``` r
counts <- data.frame(matrix(
  sample(0:100, size = 16),
  ncol = 4,
  dimnames = list(c("ENSG00000229807.12", "ENSG00000183878.12",
                    "ENSG00000239807.12", "ENSG00000259807.12"),
                  c("S1", "S2", "S2", "S4"))
))
counts
#>                    S1 S2 S2.1 S4
#> ENSG00000229807.12 19 91   59 84
#> ENSG00000183878.12 53 68   73 46
#> ENSG00000239807.12 63 67   97  0
#> ENSG00000259807.12 23 94   74 22
```

<sup>Created on 2021-07-30 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

So, if you remove the test from lines 42 through 59, your tests should all be passing! 

Let's update the function description to be only about checking whether sample identifiers match each other in metadata and counts.

**Review by kelshmo — APPROVED** [review]

**Comment by kelshmo** [comment]
Thanks @daisyhan97! One step you missed is to run `devtools::document()` that would create the documentation file for `check_mismatch` in the man/ folder. This also updates the NAMESPACE file. I ran this for you. 

I also added you as an author on the package :) Check out the package homepage.

The checks are failing due to an issue unrelated to this PR. I tested the package installs locally OK.
